### ISSUE TITLE
Rollover in RINEX output

### ISSUE THREAD

**Issue by matteocutugno** [gnss-sdr/gnss-sdr#271]
Dear all,

i'm experiencing that since GPS week Rollover happened last 6th of April RINEX output from my receiver is affected by this issue. I get year 1999 in both nav and obs file.
I use GNSS-SDR 0.0.10 in Ubuntu 16.04 acquiring signal with Nuand bladeRF x40.
How can this problem been solved?

Thank you 

MC

**Comment by acebrianjuan** [comment]
Hi @matteocutugno 

Today I have made a live capture using an RTL-SDR V3 dongle with the latest version of GNSS-SDR (version 0.0.10.git-next-cd0e7ada) and I have not found any anomaly in the dates reported in the RINEX NAV and OBS files.

As you can see in the following snippets, today's date (05/05/2019) is reported correctly.

- OBS file:

```
     3.02           OBSERVATION DATA    G                   RINEX VERSION / TYPE
G = GPS  R = GLONASS  E = GALILEO  S = GEO  M = MIXED       COMMENT             
GNSS-SDR            alvaro              20190505 115300 UTC PGM / RUN BY / DATE 
GPS OBSERVATION DATA FILE GENERATED BY GNSS-SDR             COMMENT             
GNSS-SDR VERSION 0.0.10.git-next-cd0e7ada                   COMMENT             
See https://gnss-sdr.org                                    COMMENT             
DEFAULT MARKER NAME                                         MARKER NAME         
alvaro              CTTC                                    OBSERVER / AGENCY   
GNSS-SDR            Software Receiver   0.0.10.gi           REC # / TYPE / VERS 
Antenna number      Antenna type                            ANT # / TYPE        
        0.0000        0.0000        0.0000                  APPROX POSITION XYZ 
        0.0000        0.0000        0.0000                  ANTENNA: DELTA H/E/N
G    4 C1C L1C D1C S1C                                      SYS / # / OBS TYPES 
DBHZ                                                        SIGNAL STRENGTH UNIT
  2019    05    05    11    53   17.9000000     GPS         TIME OF FIRST OBS   
                                                            END OF HEADER       
> 2019 05 05 11 53 18.0000000  0  4                                             
G25  15710592.131 7    -44700.981 7       817.978 7        45.262               
G24  16698651.901 7    137806.580 7     -2308.452 7        43.463               
G12  14283310.455 8    145960.510 8     -1084.024 8        49.171               
G14  17378638.582 6   -145864.197 6      1017.366 6        40.722               
...
```

- NAV file:

```
     3.02           N: GNSS NAV DATA    G: GPS              RINEX VERSION / TYPE
GNSS-SDR            alvaro              20190505 115300 UTC PGM / RUN BY / DATE 
GPS NAVIGATION MESSAGE FILE GENERATED BY GNSS-SDR           COMMENT             
GNSS-SDR VERSION 0.0.10.git-next-cd0e7ada                   COMMENT             
See https://gnss-sdr.org                                    COMMENT             
GPSA    .0000D+00   .0000D+00   .0000D+00   .0000D+00       IONOSPHERIC CORR    
GPSB    .0000D+00   .0000D+00   .0000D+00   .0000D+00       IONOSPHERIC CORR    
GPUT   .0000000000D+00  .000000000D+00      0 1024          TIME SYSTEM CORR    
     0     0     0     0                                    LEAP SECONDS        
                                                            END OF HEADER       
G12 2019 05 05 12 00 00  .245233066380D-03 -.341060513165D-11  .000000000000D+00
      .270000000000D+02  .265625000000D+01  .447661504041D-08  .112396095801D+01
      .502914190292D-07  .765871407930D-02  .130943953991D-05  .515364822578D+04
      .432000000000D+05 -.158324837685D-06  .164846251450D+01  .223517417908D-07
      .982716922584D+00  .367812500000D+03  .107654343088D+01 -.845178062202D-08
     -.158935191717D-09  .100000000000D+01  .102800000000D+04  .100000000000D+01
      .000000000000D+00  .000000000000D+00 -.121071934700D-07  .270000000000D+02
      .427980000000D+05  .400000000000D+01                                      
...
```

I am running Ubuntu 18.04.

I would like to learn more details about your issue to see if I can replicate it with my setup.

Can you please share your RINEX files?
What receiver configuration were you using?

Álvaro

**Comment by matteocutugno** [comment]
Hi,
first of all thank you for your reply.

I use:

bladeRF-cli version:        1.7.1-git-1da130c
libbladeRF version:         2.2.0-git-1da130c
gnss-sdr version 0.0.10.git-next-8799098
ubuntu 16.04LTS  on a  Intel® Core™ i7-8550U CPU @ 1.80GHz × 8 

My hardware configuration for the acquisition is completed by a Nuand bladerf x40 bias tee and active ublox antenna.
Here my RINEX obs and nav files:


     3.02           OBSERVATION DATA    M                   RINEX VERSION / TYPE
G = GPS  R = GLONASS  E = GALILEO  S = GEO  M = MIXED       COMMENT             
GNSS-SDR            UNKNOWN USER        20190504 113542 UTC PGM / RUN BY / DATE 
MIXED (GPS/GALILEO) OBSERVATION DATA FILE GENERATED BY GNSS-COMMENT             
GNSS-SDR VERSION 0.0.10.git-next-8799098                    COMMENT             
See https://gnss-sdr.org                                    COMMENT             
DEFAULT MARKER NAME                                         MARKER NAME         
UNKNOWN USER        CTTC                                    OBSERVER / AGENCY   
GNSS-SDR            Software Receiver   0.0.10.gi           REC # / TYPE / VERS 
Antenna number      Antenna type                            ANT # / TYPE        
        0.0000        0.0000        0.0000                  APPROX POSITION XYZ 
        0.0000        0.0000        0.0000                  ANTENNA: DELTA H/E/N
G    4 C1C L1C D1C S1C                                      SYS / # / OBS TYPES 
E    4 C1B L1B D1B S1B                                      SYS / # / OBS TYPES 
DBHZ                                                        SIGNAL STRENGTH UNIT
  1999    09    18    11    33   59.3600000     GPS         TIME OF FIRST OBS   
                                                            END OF HEADER       
> 1999 09 18 11 34 00.0000000  0  5                                             
G326169692606.780 6     20155.190 6      -786.397 6        36.860               
G256171961609.741 6   -307204.118 6       908.087 6        38.829               
G126171877341.144 7    391181.957 7     -1256.853 7        43.850               
E246170935835.404 6   -175161.310 6       858.275 6        39.222               
E336167399218.321 6     69371.836 6     -1328.531 6        39.075               
> 1999 09 18 11 34 01.0000000  0  5                                             
G326169692458.360 5     20943.136 5      -792.375 5        35.520               
G256171961778.532 6   -308109.753 6       907.362 6        38.650               
G126171877101.091 7    392442.382 7     -1260.415 7        43.823               
E246170935999.263 6   -176021.051 6       860.349 6        39.224               
E336167398965.259 6     70700.859 6     -1329.253 6        39.042               
> 1999 09 18 11 34 02.0000000  0  5                                             
G326169692304.007 6     21731.579 6      -794.623 6        36.172               
G256171961954.363 6   -309015.154 6       910.907 6        38.692               
G126171876861.361 7    393703.019 7     -1262.242 7        43.887               
E246170936159.620 6   -176880.685 6       859.757 6        39.162               
E336167398708.553 6     72030.195 6     -1329.332 6        38.919               
> 1999 09 18 11 34 03.0000000  0  5                                             
G326169692161.198 5     22520.698 5      -786.482 5        34.630               
G256171962129.336 6   -309920.152 6       905.377 6        38.899               
G126171876626.204 7    394964.038 7     -1258.924 7        43.930               
E246170936321.799 6   -177740.017 6       858.456 6        39.135               
E336167398457.600 6     73360.087 6     -1330.777 6        38.772               
> 1999 09 18 11 34 04.0000000  0  5                                             
G326169692010.955 5     23311.186 5      -793.653 5        35.491               
G256171962301.415 6   -310824.089 6       909.052 6        38.731               
G126171876386.616 7    396226.080 7     -1261.334 7        44.059               
E246170936488.088 6   -178598.437 6       857.122 6        39.119               
E336167398205.862 6     74691.128 6     -1332.407 6        38.849               
> 1999 09 18 11 34 05.0000000  0  5                                             
G326169691860.355 5     24102.758 5      -803.184 5        35.842               
G256171962472.028 6   -311727.152 6       908.131 6        38.682               
G126171876145.349 7    397488.925 7     -1262.934 7        44.030               
E246170936652.126 6   -179456.116 6       857.116 6        39.234               
E336167397951.192 6     76023.117 6     -1333.677 6        38.912               
> 1999 09 18 11 34 06.0000000  0  5                                             
G326169691705.538 6     24894.915 6      -787.651 6        39.302               
G256171962639.818 6   -312629.917 6       908.190 6        38.873               
G126171875903.295 7    398752.062 7     -1262.345 7        44.195               
E246170936816.485 6   -180313.640 6       857.335 6        39.142               
E336167397699.382 6     77355.483 6     -1332.790 6        38.792 
.....


     2.11           N: GNSS NAV DATA    M: MIXED            RINEX VERSION / TYPE
GNSS-SDR            UNKNOWN USER        20190504 113542 UTC PGM / RUN BY / DATE 
GNSS NAVIGATION MESSAGE FILE GENERATED BY GNSS-SDR          COMMENT             
GNSS-SDR VERSION 0.0.10.git-next-8799098                    COMMENT             
See https://gnss-sdr.org                                    COMMENT             
GAL     .3600D+02  -.7812D-01   .8301D-02   .0000D+00       IONOSPHERIC CORR    
GPSA    .8382D-08   .2235D-07  -.5960D-07  -.1192D-06       IONOSPHERIC CORR    
GAUT   .9313225746D-09  .000000000D+00 518400    3          TIME SYSTEM CORR    
GPGA   .0000000000D+00  .000000000D+00      0    0          TIME SYSTEM CORR    
GPUT   .0000000000D+00 -.177635684D-14 147456 1028          TIME SYSTEM CORR    
    18    18   137     7                                    LEAP SECONDS        
                                                            END OF HEADER       
G06 1999 09 18 12 00 00  .214402563870D-03 -.977706804406D-11  .000000000000D+00
      .270000000000D+02  .410000000000D+02  .400373820034D-08  .264146236287D+01
      .223144888878D-05  .158506352454D-02  .100359320641D-04  .515357416916D+04
      .561600000000D+06 -.782310962677D-07 -.248935671515D+01  .130385160446D-07
      .975127869738D+00  .193812500000D+03 -.108917245741D+01 -.769424906779D-08
      .596810573842D-09  .100000000000D+01  .102700000000D+04  .100000000000D+01
      .000000000000D+00  .000000000000D+00  .419095158577D-08  .270000000000D+02
      .559128000000D+06  .400000000000D+01                                      
G12 1999 09 18 12 00 00  .245526898652D-03 -.341060513165D-11  .000000000000D+00
      .480000000000D+02 -.300625000000D+02  .447232914760D-08  .108906190177D+01
     -.163353979588D-05  .766065088101D-02  .211410224438D-05  .515364818382D+04
      .561600000000D+06 -.154599547386D-06  .176959790121D+01  .484287738800D-07
      .982721694623D+00  .352281250000D+03  .107547118215D+01 -.841356474446D-08
     -.302512600864D-09  .100000000000D+01  .102700000000D+04  .100000000000D+01
      .000000000000D+00  .000000000000D+00 -.121071934700D-07  .480000000000D+02
      .560148000000D+06  .400000000000D+01                                      
G14 1999 09 18 12 00 00 -.818958505988D-04  .159161572810D-11  .000000000000D+00
      .600000000000D+01 -.133750000000D+02  .438161108312D-08 -.189326818702D+01
     -.622123479843D-06  .105474012671D-01  .116936862469D-04  .515370088387D+04
      .561600000000D+06  .165775418282D-06 -.313446333006D+00  .931322574615D-08
      .960866210540D+00  .149937500000D+03 -.194329550002D+01 -.772389315972D-08
     -.357872049664D-09  .100000000000D+01  .102700000000D+04  .100000000000D+01
      .000000000000D+00  .000000000000D+00 -.977888703346D-08  .600000000000D+01
      .558798000000D+06  .400000000000D+01                                      
G25 1999 09 18 12 00 00 -.752722378820D-03 -.397903932026D-11  .000000000000D+00
      .310000000000D+02 -.145312500000D+02  .481627204563D-08  .750734202476D+00
     -.655651092529D-06  .844668655191D-02  .146962702274D-05  .515371416855D+04
      .561600000000D+06 -.100582838058D-06  .170551750894D+01  .856816768646D-07
      .970169003159D+00  .355750000000D+03  .893177858523D+00 -.862607359630D-08
     -.251081887140D-09  .100000000000D+01  .102700000000D+04  .100000000000D+01
      .000000000000D+00  .000000000000D+00  .605359673500D-08  .310000000000D+02
      .560148000000D+06  .400000000000D+01                                      
G32 1999 09 18 11 59 44 -.102980528027D-03  .171667124960D-10  .000000000000D+00
      .510000000000D+02 -.254062500000D+02  .459269130403D-08 -.101236552634D+01
     -.124983489513D-05  .299883808475D-02  .113360583782D-04  .515358719063D+04
      .561584000000D+06  .409781932831D-07 -.399447588630D+00  .186264514923D-08
      .957026700118D+00  .158125000000D+03 -.251874319764D+01 -.796140305296D-08
     -.459662003910D-09  .100000000000D+01  .102700000000D+04  .100000000000D+01
      .000000000000D+00  .000000000000D+00  .465661287308D-09  .510000000000D+02
      .559818000000D+06  .400000000000D+01                                      
E24 2019 05 04 11 10 00  .609909190098D-02 -.197815097636D-10  .000000000000D+00
      .350000000000D+02  .869062500000D+02  .334513933848D-08  .522207547332D+00
      .409781932831D-05  .537866028026D-03 -.163912773132D-06  .544060123062D+04
      .558600000000D+06 -.149011611938D-07  .135696098550D+01  .167638063431D-07
      .985470409438D+00  .356562500000D+03  .101136035826D+01 -.604203738940D-08
      .210723063176D-09  .513000000000D+03  .205100000000D+04  .000000000000D+00
      .000000000000D+00  .000000000000D+00  .454019755125D-07  .507570803165D-07
      .560125000000D+06  .000000000000D+00                                      
E33 2019 05 04 10 50 00 -.470348109957D-03  .994759830064D-13  .000000000000D+00
      .330000000000D+02 -.418750000000D+01  .226580866574D-08  .154665289296D+01
     -.145286321640D-06  .337839825079D-03  .141784548759D-04  .544060880470D+04
      .557400000000D+06 -.152736902237D-06 -.285029849762D+01  .409781932831D-07
      .987070967628D+00  .494375000000D+02 -.674041911523D+00 -.503056668615D-08
      .436089593453D-09  .513000000000D+03  .205100000000D+04  .000000000000D+00
      .000000000000D+00  .000000000000D+00 -.582076609135D-08 -.651925802231D-08
      .560125000000D+06  .000000000000D+00           
..........

Here my configuration file:

[GNSS-SDR]

;######### GLOBAL OPTIONS ##################
GNSS-SDR.internal_fs_sps=2000000

;######### SIGNAL_SOURCE CONFIG ############
SignalSource.implementation=Osmosdr_Signal_Source
SignalSource.item_type=gr_complex
SignalSource.sampling_frequency=2000000
SignalSource.freq=1575420000
;# RF Gain: LNA Gain {0, 3, 6}
SignalSource.gain=6
;# IF Gain: N/A
SignalSource.rf_gain=40
;# BB Gain: RX VGA1 + VGA2 [5, 60]
SignalSource.if_gain=48
SignalSource.AGC_enabled=false
SignalSource.samples=0
SignalSource.repeat=false
SignalSource.osmosdr_args=bladerf=0 ; This line enables the bladeRF
SignalSource.enable_throttle_control=false
SignalSource.dump=false
SignalSource.dump_filename=./signal_source.dat

;######### SIGNAL_CONDITIONER CONFIG ############
SignalConditioner.implementation=Pass_Through

;######### DATA_TYPE_ADAPTER CONFIG ############
DataTypeAdapter.implementation=Pass_Through

;######### INPUT_FILTER CONFIG ############
InputFilter.implementation=Freq_Xlating_Fir_Filter
InputFilter.decimation_factor=1
InputFilter.input_item_type=gr_complex
InputFilter.output_item_type=gr_complex
InputFilter.taps_item_type=float
InputFilter.number_of_taps=5
InputFilter.number_of_bands=2
InputFilter.band1_begin=0.0
InputFilter.band1_end=0.85
InputFilter.band2_begin=0.9
InputFilter.band2_end=1.0
InputFilter.ampl1_begin=1.0
InputFilter.ampl1_end=1.0
InputFilter.ampl2_begin=0.0
InputFilter.ampl2_end=0.0
InputFilter.band1_error=1.0
InputFilter.band2_error=1.0
InputFilter.filter_type=bandpass
InputFilter.grid_density=16
InputFilter.dump=false
InputFilter.dump_filename=../data/input_filter.dat

;######### CHANNELS GLOBAL CONFIG ############
Channels_1C.count=16
Channels_1B.count=16
Channels.in_acquisition=1

Channel0.signal=1C
Channel1.signal=1C
Channel2.signal=1C
Channel3.signal=1C
Channel4.signal=1C
Channel5.signal=1C
Channel6.signal=1C
Channel7.signal=1C
Channel8.signal=1C
Channel9.signal=1C
Channel10.signal=1C
Channel11.signal=1C
Channel12.signal=1C
Channel13.signal=1C
Channel14.signal=1C
Channel15.signal=1C


Channel16.signal=1B
Channel17.signal=1B
Channel18.signal=1B
Channel19.signal=1B
Channel20.signal=1B
Channel21.signal=1B
Channel22.signal=1B
Channel23.signal=1B
Channel24.signal=1B
Channel25.signal=1B
Channel26.signal=1B
Channel27.signal=1B
Channel28.signal=1B
Channel29.signal=1B
Channel30.signal=1B
Channel31.signal=1B

;######### GPS ACQUISITION CONFIG ############
Acquisition_1C.implementation=GPS_L1_CA_PCPS_Acquisition_Fine_Doppler
Acquisition_1C.item_type=gr_complex
Acquisition_1C.coherent_integration_time_ms=1
Acquisition_1C.threshold=0.015
Acquisition_1C.doppler_max=10000
Acquisition_1C.doppler_step=250
Acquisition_1C.max_dwells=15
Acquisition_1C.repeat_satellite=true
Acquisition_1C.dump=false
Acquisition_1C.dump_filename=./acq_dump.dat

;######### GALILEO ACQUISITION CONFIG ############
Acquisition_1B.implementation=Galileo_E1_PCPS_Ambiguous_Acquisition
Acquisition_1B.item_type=gr_complex
Acquisition_1B.coherent_integration_time_ms=4
;Acquisition_1B.threshold=0.001
Acquisition_1B.pfa=0.0000002
Acquisition_1B.doppler_max=15000
Acquisition_1B.doppler_step=125
Acquisition_1B.dump=false
Acquisition_1B.dump_filename=./acq_dump.dat

;######### TRACKING GPS CONFIG ############
Tracking_1C.implementation=GPS_L1_CA_DLL_PLL_Tracking
Tracking_1C.item_type=gr_complex
Tracking_1C.pll_bw_hz=40.0;
Tracking_1C.dll_bw_hz=2.0;
Tracking_1C.order=3;
Tracking_1C.early_late_space_chips=0.5;
Tracking_1C.dump=false
Tracking_1C.dump_filename=./tracking_ch_

;######### TRACKING GALILEO CONFIG ############
Tracking_1B.implementation=Galileo_E1_DLL_PLL_VEML_Tracking
Tracking_1B.item_type=gr_complex
Tracking_1B.pll_bw_hz=15.0;
Tracking_1B.dll_bw_hz=2.0;
Tracking_1B.fll_bw_hz=10.0;
Tracking_1B.order=3;
Tracking_1B.early_late_space_chips=0.15;
Tracking_1B.very_early_late_space_chips=0.6;
Tracking_1B.dump=false
Tracking_1B.dump_filename=../data/veml_tracking_ch_

;######### TELEMETRY DECODER GPS CONFIG ############
TelemetryDecoder_1C.implementation=GPS_L1_CA_Telemetry_Decoder
TelemetryDecoder_1C.dump=false

;######### TELEMETRY DECODER GALILEO CONFIG ############
TelemetryDecoder_1B.implementation=Galileo_E1B_Telemetry_Decoder
TelemetryDecoder_1B.dump=false

;######### OBSERVABLES CONFIG ############
Observables.implementation=Hybrid_Observables
Observables.dump=true
Observables.dump_filename=./observables.dat
Observables.dump_mat=true

;######### PVT CONFIG ############
PVT.implementation=RTKLIB_PVT
PVT.positioning_mode=Single
PVT.iono_model=Broadcast
PVT.trop_model=Saastamoinen
PVT.output_rate_ms=100;
PVT.display_rate_ms=500;
PVT.dump=false
PVT.flag_rtcm_server=false
PVT.flag_rtcm_tty_port=false
PVT.rtcm_dump_devname=/dev/pts/1
PVT.dump_filename=./PVT
PVT.rinex_version=2


;######### MONITOR CONFIG ############
Monitor.enable_monitor=true
Monitor.decimation_factor=1
Monitor.client_addresses=127.0.0.1
Monitor.udp_port=1234



In my honest opinion could be some issue with PVT block because today, during other tests, gnss-sdr wasn't able to wrote RINEX files at all, obviously i got several fixes.
If you need other infos don't esitate to ask.

Thank you again 

MC

**Comment by acebrianjuan** [comment]
> gnss-sdr version 0.0.10.git-next-8799098

Ha! I just figured you are one commit behind https://github.com/gnss-sdr/gnss-sdr/commit/4bb1872e17a310e591ce32ed26031c12e3872582 _"Fix date in RINEX annotations beyond the rollover"_, which @carlesfernandez pushed on April 29th to fix the issue.

I suggest that you pull the latest changes of the [`next`](https://github.com/gnss-sdr/gnss-sdr/tree/next) branch and re-build GNSS-SDR.

The problem with the dates should be fixed.

Álvaro

**Comment by matteocutugno** [comment]
Hello, 
I upgraded and rebuilded gnss-sdr and now it writes correct RINEX. 
Thanks a lot for your help

**Comment by carlesfernandez** [comment]
There still are some annotations in RINEX files related to time which are not correct (specifically, annotations around the TOW rollover of a Saturday midnight, and also in annotations in the navigation header files related to the week number). Working on it.

**Comment by carlesfernandez** [comment]
We believe this issue is solved in new release v0.0.11. Please feel free to re-open this issue or to open a new one if the problem persists.
